### PR TITLE
Update MongoDB to use lean()

### PR DESCRIPTION
**Pull request by jason-fox**
As the mongoose docs state:

> Lean is great for high-performance, read-only cases,

### CODE REVIEW

**Comment by jason-fox** [comment]
Extraction of `lean()` from #926 but without the redis or memory cache

File: lib/services/devices/deviceRegistryMongoDB.js [hunk 1]
```diff
@@ -188,26 +188,11 @@ function listDevices(type, service, subservice, limit, offset, callback) {
     });
```
Comment (fgalan) [inline lib/services/devices/deviceRegistryMongoDB.js:188]: PR should include an entry in CHANGES_NEXT_RELEASE file describing the change.

Comment (jason-fox) [reply lib/services/devices/deviceRegistryMongoDB.js:188]: Fixed 39c7874

File: lib/services/groups/groupRegistryMongoDB.js [hunk 2]
```diff
@@ -255,36 +257,15 @@ function update(id, body, callback) {
...
-                'explicitAttrs',
-                'expressionLanguage',
-                'defaultEntityNameConjunction',
-                'ngsiVersion'
-            ];
-
-            for (let i = 0; i < attributes.length; i++) {
-                if (body[attributes[i]] !== undefined) {
-                    group[attributes[i]] = body[attributes[i]];
+            attributeList.forEach((key) => {
```
Comment (AlvaroVega) [inline lib/services/groups/groupRegistryMongoDB.js:260]: Until this `resource` was not able to be updated. With this resource could be updated. Not sure if is desirable.

Comment (jason-fox) [reply lib/services/groups/groupRegistryMongoDB.js:260]: If you really want separately maintained lists for create and update, at least bring these out as constants at the head of the file. The current position where they are buried in the middle of the file means that they are more likely to get out of sync.

It also means a new array is created and destroyed for every update.  I'm happy to re-alter the functionality here, but I'd like a clear and definite plan.

Comment (AlvaroVega) [reply lib/services/groups/groupRegistryMongoDB.js:260]: `apikey` and `resource` are keys to updae and delete a device group.
Sometime ago `apikey` and `resource` where not able to be modified. Currently `apikey` could be modified, so I don't see a reason to `resource` couldn't be modified.

**Review by AlvaroVega — APPROVED** [review]
LGTM

**Review by fgalan — APPROVED** [review]
LGTM

**Comment by AlvaroVega** [comment]
After merge this PR i've seeing raise and release MONGO alarms logs again:


time=2021-03-23T08:47:20.926Z | lvl=DEBUG | corr=24825d84-0e66-4e0f-b7f9-1530632308e5 | trans=24825d84-0e66-4e0f-b7f9-1530632308e5 | op=IoTAgentNGSI.GenericMiddlewares | from=n/a | srv=smartcity | subsrv=/ | msg=Request for path [/iot/devices] from [localhost:18082] | comp=IoTAgent
time=2021-03-23T08:47:20.926Z | lvl=DEBUG | corr=24825d84-0e66-4e0f-b7f9-1530632308e5 | trans=24825d84-0e66-4e0f-b7f9-1530632308e5 | op=IoTAgentNGSI.GenericMiddlewares | from=n/a | srv=smartcity | subsrv=/ | msg=Body:

{
    "devices": [
        {
            "device_id": "AIQ_A464",
            "entity_name": "IAQ_MA001",
            "entity_type": "IAQ",
            "static_attributes": [
                {
                    "value": "36.71749, -4.4743",
                    "type": "geo:point",
                    "name": "location"
                },
                {
                    "name": "address",
                    "type": "String",
                    "value": "Facultad"
                }
            ],
            "protocol": "IoTA-UL"
        }
    ]
}

 | comp=IoTAgent
time=2021-03-23T08:47:20.928Z | lvl=DEBUG | corr=24825d84-0e66-4e0f-b7f9-1530632308e5 | trans=24825d84-0e66-4e0f-b7f9-1530632308e5 | op=IoTAgentNGSI.DeviceProvisioning | from=n/a | srv=smartcity | subsrv=/ | msg=Handling device provisioning request. | comp=IoTAgent
time=2021-03-23T08:47:20.929Z | lvl=DEBUG | corr=24825d84-0e66-4e0f-b7f9-1530632308e5 | trans=24825d84-0e66-4e0f-b7f9-1530632308e5 | op=IoTAgentNGSI.MongoDBGroupRegister | from=n/a | srv=n/a | subsrv=n/a | msg=Looking for group params ["service","subservice","type","apikey"] with queryObj {"service":"smartcity","subservice":"/","type":"IAQ"} | comp=IoTAgent
time=2021-03-23T08:47:20.934Z | lvl=DEBUG | corr=94af403e-9590-4093-8d8a-5207085f8eaf | trans=94af403e-9590-4093-8d8a-5207085f8eaf | op=IoTAgentNGSI.MongoDBGroupRegister | from=n/a | srv=n/a | subsrv=n/a | msg=Device group for fields [["service","subservice","type","apikey"]] not found: [{"service":"smartcity","subservice":"/","type":"IAQ"}] | comp=IoTAgent
time=2021-03-23T08:47:20.934Z | lvl=DEBUG | corr=94af403e-9590-4093-8d8a-5207085f8eaf | trans=94af403e-9590-4093-8d8a-5207085f8eaf | op=IoTAgentNGSI.BidirectionalPlugin | from=n/a | srv=n/a | subsrv=n/a | msg=Extracting attribute list | comp=IoTAgent
time=2021-03-23T08:47:20.934Z | lvl=DEBUG | corr=94af403e-9590-4093-8d8a-5207085f8eaf | trans=94af403e-9590-4093-8d8a-5207085f8eaf | op=IoTAgentNGSI.BidirectionalPlugin | from=n/a | srv=n/a | subsrv=n/a | msg=Sending bidirectionality subscriptions for device [AIQ_A464] | comp=IoTAgent
time=2021-03-23T08:47:20.935Z | lvl=DEBUG | corr=94af403e-9590-4093-8d8a-5207085f8eaf | trans=94af403e-9590-4093-8d8a-5207085f8eaf | op=IoTAgentNGSI.ContextServer | from=n/a | srv=n/a | subsrv=n/a | msg=Looking for bindings for the function [deviceProvisioningHandler] and protocol [null] | comp=IoTAgent
time=2021-03-23T08:47:20.935Z | lvl=DEBUG | corr=94af403e-9590-4093-8d8a-5207085f8eaf | trans=94af403e-9590-4093-8d8a-5207085f8eaf | op=IoTAgentNGSI.ContextServer | from=n/a | srv=n/a | subsrv=n/a | msg=Binding found for function [deviceProvisioningHandler] and protocol [null] | comp=IoTAgent
time=2021-03-23T08:47:20.936Z | lvl=DEBUG | corr=94af403e-9590-4093-8d8a-5207085f8eaf | trans=94af403e-9590-4093-8d8a-5207085f8eaf | op=IoTAgentNGSI.MongoDBDeviceRegister | from=n/a | srv=smartcity | subsrv=/ | msg=Looking for device with id [AIQ_A464]. | comp=IoTAgent
time=2021-03-23T08:47:20.939Z | lvl=DEBUG | corr=94af403e-9590-4093-8d8a-5207085f8eaf | trans=94af403e-9590-4093-8d8a-5207085f8eaf | op=IoTAgentNGSI.MongoDBDeviceRegister | from=n/a | srv=smartcity | subsrv=/ | msg=Device [AIQ_A464] not found. | comp=IoTAgent
**time=2021-03-23T08:47:20.939Z | lvl=ERROR | corr=94af403e-9590-4093-8d8a-5207085f8eaf | trans=94af403e-9590-4093-8d8a-5207085f8eaf | op=IoTAgentNGSI.Alarms | from=n/a | srv=n/a | subsrv=n/a | msg=Raising [MONGO-ALARM]: {"name":"DEVICE_NOT_FOUND","message":"No device was found with id:AIQ_A464","code":404} | comp=IoTAgent**
time=2021-03-23T08:47:20.939Z | lvl=DEBUG | corr=94af403e-9590-4093-8d8a-5207085f8eaf | trans=94af403e-9590-4093-8d8a-5207085f8eaf | op=IoTAgentNGSI.MongoDBGroupRegister | from=n/a | srv=n/a | subsrv=n/a | msg=Looking for group params ["service","subservice","type","apikey"] with queryObj {"service":"smartcity","subservice":"/","type":"IAQ"} | comp=IoTAgent
time=2021-03-23T08:47:20.942Z | lvl=DEBUG | corr=94af403e-9590-4093-8d8a-5207085f8eaf | trans=94af403e-9590-4093-8d8a-5207085f8eaf | op=IoTAgentNGSI.MongoDBGroupRegister | from=n/a | srv=n/a | subsrv=n/a | msg=Device group for fields [["service","subservice","type","apikey"]] not found: [{"service":"smartcity","subservice":"/","type":"IAQ"}] | comp=IoTAgent
time=2021-03-23T08:47:20.943Z | lvl=DEBUG | corr=94af403e-9590-4093-8d8a-5207085f8eaf | trans=94af403e-9590-4093-8d8a-5207085f8eaf | op=IoTAgentNGSI.DeviceService | from=n/a | srv=n/a | subsrv=n/a | msg=deviceData before merge with conf: {"id":"AIQ_A464","type":"IAQ","name":"IAQ_MA001","service":"smartcity","subservice":"/","staticAttributes":[{"value":"36.71749, -4.4743","type":"geo:point","name":"location"},{"name":"address","type":"String","value":"Facultad de Medicina"}],"protocol":"IoTA-UL","transport":"HTTP","internalId":null,"explicitAttrs":false,"subscriptions":[],"polling":true} | comp=IoTAgent
time=2021-03-23T08:47:20.944Z | lvl=DEBUG | corr=94af403e-9590-4093-8d8a-5207085f8eaf | trans=94af403e-9590-4093-8d8a-5207085f8eaf | op=IoTAgentNGSI.DeviceService | from=n/a | srv=n/a | subsrv=n/a | msg=deviceData after merge with conf: {"id":"AIQ_A464","type":"IAQ","name":"IAQ_MA001","service":"smartcity","subservice":"/","active":null,"staticAttributes":[{"value":"36.71749, -4.4743","type":"geo:point","name":"location"},{"name":"address","type":"String","value":"Facultad de Medicina"}],"lazy":null,"commands":[],"protocol":"IoTA-UL","transport":"HTTP","internalId":null,"explicitAttrs":false,"subscriptions":[],"polling":true} | comp=IoTAgent
time=2021-03-23T08:47:20.944Z | lvl=DEBUG | corr=94af403e-9590-4093-8d8a-5207085f8eaf | trans=94af403e-9590-4093-8d8a-5207085f8eaf | op=IoTAgentNGSI.DeviceService | from=n/a | srv=n/a | subsrv=n/a | msg=Registering device into NGSI Service:
{
    "id": "AIQ_A464",
    "type": "IAQ",
    "name": "IAQ_MA001",
    "service": "smartcity",
    "subservice": "/",
    "active": null,
    "staticAttributes": [
        {
            "value": "36.71749, -4.4743",
            "type": "geo:point",
            "name": "location"
        },
        {
            "name": "address",
            "type": "String",
            "value": "Facultad"
        }
    ],
    "lazy": null,
    "commands": [],
    "protocol": "IoTA-UL",
    "transport": "HTTP",
    "internalId": null,
    "explicitAttrs": false,
    "subscriptions": [],
    "polling": true
} | comp=IoTAgent
time=2021-03-23T08:47:20.945Z | lvl=DEBUG | corr=94af403e-9590-4093-8d8a-5207085f8eaf | trans=94af403e-9590-4093-8d8a-5207085f8eaf | op=IoTAgentNGSI.Registration | from=n/a | srv=n/a | subsrv=n/a | msg=Registration with Context Provider is not needed. Device without lazy atts or commands | comp=IoTAgent
time=2021-03-23T08:47:20.946Z | lvl=DEBUG | corr=94af403e-9590-4093-8d8a-5207085f8eaf | trans=94af403e-9590-4093-8d8a-5207085f8eaf | op=IoTAgentNGSI.Devices-v2 | from=n/a | srv=n/a | subsrv=n/a | msg=deviceData: {"id":"AIQ_A464","type":"IAQ","name":"IAQ_MA001","service":"smartcity","subservice":"/","active":null,"staticAttributes":[{"value":"36.71749, -4.4743","type":"geo:point","name":"location"},{"name":"address","type":"String","value":"Facultad de Medicina"}],"lazy":null,"commands":[],"protocol":"IoTA-UL","transport":"HTTP","internalId":null,"explicitAttrs":false,"subscriptions":[],"polling":true} | comp=IoTAgent
time=2021-03-23T08:47:20.947Z | lvl=DEBUG | corr=94af403e-9590-4093-8d8a-5207085f8eaf | trans=94af403e-9590-4093-8d8a-5207085f8eaf | op=IoTAgentNGSI.Devices-v2 | from=n/a | srv=n/a | subsrv=n/a | msg=config.timestamp undefined true | comp=IoTAgent
time=2021-03-23T08:47:20.949Z | lvl=DEBUG | corr=94af403e-9590-4093-8d8a-5207085f8eaf | trans=94af403e-9590-4093-8d8a-5207085f8eaf | op=IoTAgentNGSI.Devices-v2 | from=n/a | srv=n/a | subsrv=n/a | msg=Creating initial entity in the Context Broker:
 {
    "url": "http://iot-orion:1026/v2/entities?options=upsert",
    "method": "POST",
    "json": {
        "id": "IAQ_MA001",
        "type": "IAQ",
        "location": {
            "type": "geo:json",
            "value": {
                "type": "Point",
                "coordinates": [
                    36.71749,
                    -4.4743
                ]
            }
        },
        "address": {
            "type": "String",
            "value": "Facultad de Medicina"
        },
        "TimeInstant": {
            "type": "DateTime",
            "value": "2021-03-23T08:47:20.948Z"
        }
    },
    "headers": {
        "fiware-service": "smartcity",
        "fiware-servicepath": "/",
        "fiware-correlator": "94af403e-9590-4093-8d8a-5207085f8eaf"
    }
} | comp=IoTAgent
time=2021-03-23T08:47:20.949Z | lvl=DEBUG | corr=94af403e-9590-4093-8d8a-5207085f8eaf | trans=94af403e-9590-4093-8d8a-5207085f8eaf | op=IoTAgentNGSI.RestUtils | from=n/a | srv=n/a | subsrv=n/a | msg=executeWithSecurity | comp=IoTAgent
time=2021-03-23T08:47:20.949Z | lvl=DEBUG | corr=94af403e-9590-4093-8d8a-5207085f8eaf | trans=94af403e-9590-4093-8d8a-5207085f8eaf | op=IoTAgentNGSI.MongoDBGroupRegister | from=n/a | srv=n/a | subsrv=n/a | msg=Looking for group params ["type"] with queryObj {"type":"IAQ"} | comp=IoTAgent
time=2021-03-23T08:47:20.952Z | lvl=DEBUG | corr=94af403e-9590-4093-8d8a-5207085f8eaf | trans=94af403e-9590-4093-8d8a-5207085f8eaf | op=IoTAgentNGSI.MongoDBGroupRegister | from=n/a | srv=smartcity | subsrv=/a | msg=Device group data found: {"_id":"60584888d434e17f675b761a","commands":[],"staticAttributes":[],"attributes":[],"resource":"/iot/d","apikey":"xvp8pp1yrsexyaac92z1di3my","type":"IAQ","service":"smartcity","subservice":"/a","description":"ul-iaq","timestamp":true} | comp=IoTAgent
**time=2021-03-23T08:47:20.953Z | lvl=ERROR | corr=94af403e-9590-4093-8d8a-5207085f8eaf | trans=94af403e-9590-4093-8d8a-5207085f8eaf | op=IoTAgentNGSI.Alarms | from=n/a | srv=n/a | subsrv=n/a | msg=Releasing [MONGO-ALARM] | comp=IoTAgent**
time=2021-03-23T08:47:20.971Z | lvl=DEBUG | corr=94af403e-9590-4093-8d8a-5207085f8eaf | trans=94af403e-9590-4093-8d8a-5207085f8eaf | op=IoTAgentNGSI.Devices-v2 | from=n/a | srv=n/a | subsrv=n/a | msg=Initial entity created successfully. | comp=IoTAgent
time=2021-03-23T08:47:20.989Z | lvl=DEBUG | corr=94af403e-9590-4093-8d8a-5207085f8eaf | trans=94af403e-9590-4093-8d8a-5207085f8eaf | op=IoTAgentNGSI.MongoDBDeviceRegister | from=n/a | srv=smartcity | subsrv=/ | msg=Storing device with id [AIQ_A464] and type [IAQ] | comp=IoTAgent
time=2021-03-23T08:47:21.020Z | lvl=DEBUG | corr=94af403e-9590-4093-8d8a-5207085f8eaf | trans=94af403e-9590-4093-8d8a-5207085f8eaf | op=IoTAgentNGSI.DeviceProvisioning | from=n/a | srv=n/a | subsrv=n/a | msg=Device provisioning request succeeded | comp=IoTAgent
time=2021-03-23T08:47:21.022Z | lvl=DEBUG | corr=24825d84-0e66-4e0f-b7f9-1530632308e5 | trans=24825d84-0e66-4e0f-b7f9-1530632308e5 | op=IoTAgentNGSI.DomainControl | from=n/a | srv=smartcity | subsrv=/ | msg=response-time: 112 | comp=IoTAgent

**Pull request merged**